7_drivers_munge/src/GCM_driver_utils.R

```diff
@@ -26,34 +10,170 @@ split_lake_centroids <- function(centroids_sf_rds) {
+      sf::st_transform(crs = sf::st_crs(grid_cells))
+
+    # build plot
+    query_plot <- ggplot() +
+      geom_sf(data=wi_sf, fill=NA, color='grey70') +
+      geom_sf(data=grid_cells, color='grey50', fill=NA, size=0.5) +
+      geom_sf(data=grid_tiles, color='dodgerblue3', fill=NA) +
+      geom_sf(data=cells_w_lakes, color='firebrick3', fill=NA) +
+      geom_sf(data=selected_tile, color='darkturquoise', fill=NA) +
+      geom_sf(data=lake_centroids, color='black', size=0.5) +
+      coord_sf() + theme_void()
+
+    # save file
+    out_file <- gsub('XX', tile_id, out_file_template)
+    ggsave(out_file, query_plot, width=10, height=8, dpi=300)
+
+    return(out_file)
+  } else {
+    # MUST BE A BETTER WORKAROUND?
+    # Build placeholder out_file name
+    out_file <- gsub('XX', 'NULL', out_file_template)
+
+    # write empty file
+    query_plot_empty <- ggplot()
+    ggsave(out_file, query_plot_empty)
+
+    return(out_file)
+  }
+}
+
 # Convert an sf object into a geoknife::simplegeom, so that
 # it can be used in the geoknife query. `geoknife` only works
 # with `sp` objects but not SpatialPoints at the moment, so
 # need to convert these to a data.frame.
 sf_pts_to_simplegeom <- function(sf_obj) {
-  sf_obj %>%
-    st_coordinates() %>%
-    t() %>% as.data.frame() %>%
-    simplegeom()
+    sf_obj %>%
+      st_coordinates() %>%
+      t() %>% as.data.frame() %>%
+      simplegeom()
 }
 
 # Set up a download file to get the raw GCM data down.
 # This function accepts each of the query parameters for the
 # geoknife job as an argument. Currently missing the "knife" parameter.
-download_gcm_data <- function(out_file, query_geom, query_url, query_vars,
+download_gcm_data <- function(out_file_template, query_geom, query_url_template, gcm_name, query_vars,
                               query_dates, query_knife = NULL) {
-  gcm_job <- geoknife(
-    stencil = query_geom,
-    fabric = webdata(
-      url = query_url,
-      variables = query_vars,
-      times = query_dates
+  #IF TILE CONTAINS CELLS THAT CONTAIN LAKES
+  if (nrow(query_geom) > 0) {
+
+    # Reproject query cell centroids to WGS84
+    query_geom_WGS84 <- project_to_crs(query_geom, crs = 4326)
+
+    # convert grid cell centroids into geoknife-friendly format
+    query_simplegeom <- sf_pts_to_simplegeom(query_geom_WGS84)
+
+    # Build query_url
+    query_url <- gsub('GCMname', gcm_name, query_url_template)
+
+    # construct and submit query
+    gcm_job <- geoknife(
+      stencil = query_simplegeom,
+      fabric = webdata(
+        url = query_url,
+        variables = query_vars,
+        times = query_dates
+      )
     )
-  )
+    wait(gcm_job)
+    my_data <- result(gcm_job)
+
+    # Build out_file name
+    out_file <- gsub('GCMname', gcm_name, out_file_template)
+    out_file <- gsub('XX', query_geom$tile_no[1], out_file)
+
+    # write file
+    arrow::write_feather(my_data, out_file)
+
+    return(out_file)
+  } else {
+    # MUST BE A BETTER WORKAROUND?
```

Comment:
Can you instead throw an error in this function when `nrow(query_geom) == 0` (I'd recommend throwing the error immediately at the top of the function) and instead filter out all targets that don't have any geoms (i.e., the empty ones) prior to branching?

Reply:
That's what @lindsayplatt and I were hoping to do, but we're running into some complications with `targets` and `sf` objects, where we can't stop branching over a `sf` object without `targets` throwing an error. We'll keep trying to find a better approach than this -- this is meant as a hacky workaround for now.